### PR TITLE
add TP and gptj model support

### PR DESCRIPTION
**Pull request by ClarkChin08**
1. add TP_1D algo
2. add parallel_context for broadcast/reduce
3. support all data type
4. support gptj model

## Type of Change

feature

## Dependency Change?

any oneCCL for distributed communication

### CODE REVIEW

**Comment by hshen14** [comment]
I heard from Pujiang that 2S AutoTP does not reply on oneCCL. Does it work for us?

**Comment by hshen14** [comment]
> > 
> 
> yes, the broadcast and all_reduce can use other library or implemented by themselves , but now i use oneccl to build the parallel_context.

I meant through memory directly instead of communication libraries on 2 sockets :)

**Comment by hshen14** [comment]
> > > > 
> > > 
> > > 
> > > yes, the broadcast and all_reduce can use other library or implemented by themselves , but now i use oneccl to build the parallel_context.
> > 
> > 
> > I meant through memory directly instead of communication libraries on 2 sockets :)
> 
> oh, have not thought on that. maybe we can do that, but there are two time-consuming issue
> 
> 1. we also need to sync the data between 2 sockets, because the calculation is not always finished at same time even we use same memory buffer(as the memory will have read/write, not like the weights which are read-only).
> 2. even we use shared-memory like the weights(mmap), the cross-socket memory  movement also cost time(the test data shows, if i allocate the memory in socket-0 and use it through shared-memory in socket-1, the gptj-q4-0 latency will drop from 33ms -> 53ms). all-reduce also have the memory-movement between socket but the size is quite small( the Q weight may 4096* 4096, but the results shape maybe 32* 4096).
>    even though , this 2 reasons have relationship with each other. that means , in TP if we want to write/read at the same time we need to sync the data( easy to say, we have to wait the slow one to finished the calculation), but if we have weight movement through sockets , many time wasted here ,and the perf will be not good.
>    That's why we keep weight allocated on each sockets and each part only have half piece of the weight. Then after the matmul finished the calculation, the two sockets communicate ONLY on the output which is quite small shape(all-reduce), we may implement more efficient way on the communication ops.

sync is fine. I thought the tp through memory is what Yu suggested at the beginning and Pujiang demonstrated it works later?

File: intel_extension_for_transformers/llm/runtime/graph/tensor_parallelism.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+cmake -DNE_TP=ON .. 
+make -j
+
+```
+
+## Examples
+
+We can config the `mpirun` to start parallel programs. Here is an example about running tensor pallelsim on 2 sockets in CPU.
+```shell
+mpirun -np 2 -bind-to=socket ./build/bin/main_gptj -m gptj_q4_0.bin --seed 1234 -t 56 -c 68 -n 32 -p "Once upon a time, there existed a little girl, who liked to have adventures. She wanted to go to places and meet new people, and have fun." --no_mmap
```
Comment (kevinintel) [inline intel_extension_for_transformers/llm/runtime/graph/tensor_parallelism.md:56]: add how to generate q4_0 model.
it's better to show the support matrix like model, kernel (q4_0 or q4_j, q8_0,...)
explain why use no_mmap

**Review by kevinintel — APPROVED** [review]
LGTM

File: intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h [hunk 2]
```diff
@@ -139,6 +165,19 @@ struct model_load_tensor {
       case SPLIT_BY_ROWS:
         ne = {first_shard.ne[0], checked_mul<uint32_t>(first_shard.ne[1], n_shards)};
         break;
+#ifdef NE_TP_MODEL
+      case TP_1D_ROW:
+	MODEL_ASSERT(first_shard.ne.size() > 1);
+        ne = {first_shard.ne[0], first_shard.ne[1] / world_size};
```
Comment (luoyu-intel) [inline intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:171]: is there a limitation that ne[1] must be aligned with world_size?

Comment (luoyu-intel) [reply intel_extension_for_transformers/llm/runtime/graph/models/model_utils/model_files.h:171]: yes, you can add one assert on it.

**Review by luoyu-intel — APPROVED** [review]
OK

**Comment by airMeng** [comment]
hi, any performance data now?

**Comment by hshen14** [comment]
Can you also add the document to show how to enable TP? Did you test 2 SPR nodes (4 sockets)?

File: intel_extension_for_transformers/llm/runtime/graph/models/gptj/gptj.cpp [hunk 3]
```diff
@@ -122,6 +122,22 @@ static bool gptj_model_eval_internal(model_context& lctx, const model_token* tok
...
+  int32_t rank = get_tp_rank(p_ctx);
+  enable_tp = world_size > 1 ? true : false;
+
+  // int32_t world_size = 2;
+  // int32_t rank = 0;
+  // bool enable_tp =true;
+  if (enable_tp) {
+    // need to broadcast the ids
+    broadcast(p_ctx, (float*)embd->data,  N * batch_size * ne_element_size(embd));
+  }
```
Comment (DDEle) [inline intel_extension_for_transformers/llm/runtime/graph/models/gptj/gptj.cpp:138]: Is it necessary to broadcast the embd data? I think all ranks should have the same embed upon here.

**Review by DDEle — COMMENTED** [review]
I Tried to learn MPI and TP by following current TP implementation...

**Pull request merged**